test/python/test_instructions.py

```diff
@@ -0,0 +1,353 @@
+            test_hwm),
+        navigate=lambda: navigator.navigate(instructions))
+
+    received_main_hwm = client.get_main_hwm()
+
+    assert received_main_hwm == main_hwm, \
+        f"Expected main hmw {main_hwm} but got {received_main_hwm}"
+
+
+@pytest.mark.parametrize("account", [DEFAULT_ACCOUNT])
+def test_get_all_hwm(
+        account: Account,
+        client: TezosClient,
+        firmware: Firmware,
+        navigator: Navigator) -> None:
+    """Test the QUERY_ALL_HWM instruction."""
+
+    main_chain_id: int = 0
+    main_hwm = Hwm(0)
+    test_hwm = Hwm(0)
+
+    instructions = get_setup_app_context_instructions(firmware)
+
+    send_and_navigate(
+        send=lambda: client.setup_app_context(
+            account,
+            main_chain_id,
+            main_hwm,
+            test_hwm),
+        navigate=lambda: navigator.navigate(instructions))
+
+    received_main_chain_id, received_main_hwm, received_test_hwm = client.get_all_hwm()
+
+    assert received_main_chain_id == main_chain_id, \
+        f"Expected main chain id {main_chain_id} but got {received_main_chain_id}"
+
+    assert received_main_hwm == main_hwm, \
+        f"Expected main hmw {main_hwm} but got {received_main_hwm}"
+
+    assert received_test_hwm == test_hwm, \
+        f"Expected test hmw {test_hwm} but got {received_test_hwm}"
+
+
+@pytest.mark.parametrize("account", [DEFAULT_ACCOUNT])
+@pytest.mark.parametrize("with_hash", [False, True])
+def test_sign_message(
+        account: Account,
+        with_hash: bool,
+        client: TezosClient,
+        firmware: Firmware,
+        navigator: Navigator) -> None:
+    """Test the SIGN(_WITH_HASH) instruction."""
+
+    main_chain_id: int = 0
+    main_hwm = Hwm(0)
+    test_hwm = Hwm(0)
+
+    instructions = get_setup_app_context_instructions(firmware)
+
+    send_and_navigate(
+        send=lambda: client.setup_app_context(
+            account,
+            main_chain_id,
+            main_hwm,
+            test_hwm),
+        navigate=lambda: navigator.navigate(instructions))
+
+    tx = RawMessage("0270000007000000000000000000000000000000000000000000000000000000000000000014000000000003000000010000000000000000000000000000000000000000000000000000000000000000")
+
+    if with_hash:
+        signature = client.sign_message(account, tx)
+        account.check_signature(signature, bytes(tx))
+    else:
+        tx_hash, signature = client.sign_message_with_hash(account, tx)
+        assert tx_hash == tx.hash, \
+            f"Expected hash {tx.hash.hex()} but got {tx_hash.hex()}"
+        account.check_signature(signature, bytes(tx))
+
+
+@pytest.mark.parametrize("account", [DEFAULT_ACCOUNT])
+def test_hmac(
+        account: Account,
+        client: TezosClient) -> None:
+    """Test the HMAC instruction."""
+
+    message = bytes.fromhex("0123456789abcdef")
+    data = client.hmac(account, message)
+    hmac = bytes.fromhex("dde436940ab1602029a49dc77e6263b633fa3567c4d8479820d4f77072369ac1")
```

Comment:
A comment explaining the test would be good. How did you generate this hex.

Reply:
I don't know yet how to check properly the `HMAC` instruction.
The value called `hmac` has been generated by the app itself.
It will help prevent regression.
I will try in a future MR to check the `HMAC` properly : https://github.com/trilitech/ledger-app-tezos-baking/issues/7

Reply:
Ok. sounds good.